client/finality-grandpa-warp-sync/src/proof.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use codec::{Decode, Encode};
+
+use sc_finality_grandpa::{AuthoritySetChanges, GrandpaJustification};
+use sp_blockchain::Backend as BlockchainBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, NumberFor},
+};
+
+use crate::HandleRequestError;
+
+/// The maximum number of authority set change proofs to include in a single warp sync proof.
+const MAX_CHANGES_PER_WARP_SYNC_PROOF: usize = 256;
+
+#[derive(Decode, Encode)]
+pub struct AuthoritySetChangeProof<Block: BlockT> {
+	/// must contain the authorities for the following set
+	pub header: Block::Header,
+	/// this justification proves finality of the header above
+	/// it must be validated against the authorities of the previous set.
+	pub justification: GrandpaJustification<Block>,
+}
+
+#[derive(Decode, Encode)]
+pub struct WarpSyncProof<Block: BlockT> {
+	proofs: Vec<AuthoritySetChangeProof<Block>>,
+}
+
+impl<Block: BlockT> WarpSyncProof<Block> {
+	pub fn generate<Backend>(
+		backend: &Backend,
+		begin: Block::Hash,
+		set_changes: &AuthoritySetChanges<NumberFor<Block>>,
+	) -> Result<WarpSyncProof<Block>, HandleRequestError>
+	where
+		Backend: BlockchainBackend<Block>,
+	{
+		// TODO: cache best response (i.e. the one with lowest begin_number)
+
+		let begin_number = backend
+			.block_number_from_id(&BlockId::Hash(begin))?
+			.ok_or_else(|| HandleRequestError::InvalidRequest("Missing start block".to_string()))?;
+
+		if begin_number > backend.info().finalized_number {
+			return Err(HandleRequestError::InvalidRequest(
+				"Start block is not finalized".to_string(),
+			));
+		}
+
+		let canon_hash = backend.hash(begin_number)?.expect(
+			"begin number is lower than finalized number; \
+			 all blocks below finalized number must have been imported; \
+			 qed.",
+		);
+
+		if canon_hash != begin {
+			return Err(HandleRequestError::InvalidRequest(
+				"Start block is not in the finalized chain".to_string(),
+			));
+		}
+
+		let mut proofs = Vec::new();
+
+		for (_, last_block) in set_changes.iter() {
+			if proofs.len() >= MAX_CHANGES_PER_WARP_SYNC_PROOF {
+				break;
+			}
+
+			if *last_block <= begin_number {
+				continue;
+			}
```

Comment:
This is a performance nitpick, but could we binary-search the set changes to find the correct starting point of the iterator? I don't like that we're iterating over the whole set every time.